File: solutions/cpp/food-chain/1/food_chain.cpp
#include "food_chain.h"

namespace food_chain {
    const auto song = std::array<std::string, 8> {
        "I know an old lady who swallowed a fly.\nI don't know why she swallowed the fly. Perhaps she'll die.\n",
        "I know an old lady who swallowed a spider.\nIt wriggled and jiggled and tickled inside her.\nShe swallowed the spider to catch the fly.\nI don't know why she swallowed the fly. Perhaps she'll die.\n",
        "I know an old lady who swallowed a bird.\nHow absurd to swallow a bird!\nShe swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\nShe swallowed the spider to catch the fly.\nI don't know why she swallowed the fly. Perhaps she'll die.\n",
        "I know an old lady who swallowed a cat.\nImagine that, to swallow a cat!\nShe swallowed the cat to catch the bird.\nShe swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\nShe swallowed the spider to catch the fly.\nI don't know why she swallowed the fly. Perhaps she'll die.\n",
        "I know an old lady who swallowed a dog.\nWhat a hog, to swallow a dog!\nShe swallowed the dog to catch the cat.\nShe swallowed the cat to catch the bird.\nShe swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\nShe swallowed the spider to catch the fly.\nI don't know why she swallowed the fly. Perhaps she'll die.\n",
        "I know an old lady who swallowed a goat.\nJust opened her throat and swallowed a goat!\nShe swallowed the goat to catch the dog.\nShe swallowed the dog to catch the cat.\nShe swallowed the cat to catch the bird.\nShe swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\nShe swallowed the spider to catch the fly.\nI don't know why she swallowed the fly. Perhaps she'll die.\n",
        "I know an old lady who swallowed a cow.\nI don't know how she swallowed a cow!\nShe swallowed the cow to catch the goat.\nShe swallowed the goat to catch the dog.\nShe swallowed the dog to catch the cat.\nShe swallowed the cat to catch the bird.\nShe swallowed the bird to catch the spider that wriggled and jiggled and tickled inside her.\nShe swallowed the spider to catch the fly.\nI don't know why she swallowed the fly. Perhaps she'll die.\n",
        "I know an old lady who swallowed a horse.\nShe's dead, of course!\n"
    };

    auto verse(uint8_t number) -> std::string {
        return song.at(number - 1);
    }

    auto verses(uint8_t start, uint8_t end) -> std::string {
        return std::accumulate(song.cbegin() + start - 1, song.cbegin() + end, std::string{}, 
            [] (auto acc, auto ch) {
                return acc + ch + "\n";
            });
    }

    auto sing() -> std::string {
        return std::accumulate(song.cbegin(), song.cend(), std::string{},
            [] (auto acc, auto ch) {
                return acc + ch + "\n";
            });
    }
}  // namespace food_chain

File: solutions/javascript/food-chain/1/food-chain.js
export class Song {
  #song = (() => {
   const characters = [
      {
        character: "fly",
        entrance: "",
      },
      {
        character: "spider",
        entrance: "It wriggled and jiggled and tickled inside her.\n",
        action: " that wriggled and jiggled and tickled inside her"
      },
      {
        character: "bird",
        entrance: "How absurd to swallow a bird!\n"
      },
      {
        character: "cat",
        entrance: "Imagine that, to swallow a cat!\n"
      },
      {
        character: "dog",
        entrance: "What a hog, to swallow a dog!\n"
      },
      {
        character: "goat",
        entrance: "Just opened her throat and swallowed a goat!\n"
      },
      {
        character: "cow",
        entrance: "I don't know how she swallowed a cow!\n"
      }
   ];
   const verses = characters.reduce((acc, protagonist, idx) => {
     const verse = `I know an old lady who swallowed a ${protagonist.character}.\n` +
                   `${protagonist.entrance}` +
                   characters.slice(0, idx).reduceRight((catches, coadjuvant) => {
                     const catched = coadjuvant.character + (coadjuvant.action ?? "");
                     const phrase = `She swallowed the ${protagonist.character} to catch the ${catched}.\n`;
                     protagonist = coadjuvant;
                     return catches + phrase;
                   }, "") +
                   `I don't know why she swallowed the fly. ` +
                   `Perhaps she'll die.\n`;
     return [...acc, verse];
   }, []);
   const finale = {
     character: "horse",
     line: "She's dead, of course!"
   };       
   return verses.concat( 
     `I know an old lady who swallowed a ${finale.character}.\n` +
     `${finale.line}\n`
   );    
  })();
  
  verse = (idx) => this.#song.at(idx - 1);

  verses = (start, end) => this.#song.slice(start - 1, end).reduce((acc, verse) => acc + verse + "\n", "");
}
